docs/burger-discounts/index.md

```diff
@@ -0,0 +1,545 @@
+constructor, are turned into JS objects. OCaml arrays, like the one that
+contains all the items, are turned into JS arrays.
+
+::: warning
+
+Variant constructors in the runtime don't always have the `TAG` field. That
+field only appears when there's more than one variant constructor with an
+argument. See [Data types and runtime
+representation](https://melange.re/v3.0.0/communicate-with-javascript.html#data-types-and-runtime-representation)
+for more details.
+
+:::
+
+## Arrays are mutable
+
+The "Input array isn't changed" unit test fails because arrays in OCaml are
+mutable (just as in JavaScript) and the `Discount.getFreeBurger` function
+mutates its array argument. The easiest way to fix this is to swap the order of
+`Js.Array.sortInPlaceWith` and `Js.Array.filter` invocations:
+
+<<< Discount.re#swap-function-order
+
+Although sorting still happens in-place, the array being sorted is a new one
+created by `Js.Array.filter` (the array containing only burgers), not the
+original input array.
+
+## Runtime representation of `option`
+
+We need to add one more test to check that `Discount.getFreeBurger` works when
+there are more than two burgers:
+
+<<< DiscountTests.re#three-burgers
+
+This test fails, with the key part of the output being:
+
+```diff
++      Expected values to be strictly equal:
++      + actual - expected
++
++      + undefined
++      - 15.15
+```
+
+Recall that `Discount.getFreeBurger` has the return type `option(float)`.
+This is how Melange maps `option(float)` values to the JavaScript runtime[^3]:
+
+| OCaml source | JavaScript runtime |
+|--------------|--------------------|
+| `None` | `undefined` |
+| `Some(15.15)` | `15.15` |
+
+So Node test runner is basically telling you that `None` was returned, but
+`Some(15.15)` was expected.
+
+## Pattern matching on arrays
+
+The test is failing because the current "success" pattern match only accounts
+for a two-element array:
+
+```reason{2}
+switch (burgers) {
+| [|Burger(_), Burger(cheaperBurger)|] =>
+  Some(Item.Burger.toPrice(cheaperBurger))
+| _ => None
+```
+
+OCaml [only allows you to pattern match on arrays of fixed
+length](https://v2.ocaml.org/manual/patterns.html#sss:pat-array), so to fix
+this, we must instead match on a tuple of the first and second elements of the
+array:
+
+<<< Discount.re#match-on-tuple
+
+## Array access is unsafe
+
+The first and second tests now fail due to `index out of bounds` errors (since
+they work on arrays of length 0 and 1, respectively). Array access in OCaml is
+unsafe by default, so the simplest fix is to check the length of the array
+before using the switch expression:
+
+<<< Discount.re#check-array-length{1-2}
+
+As an extra precaution, you might want to catch the exception inside the switch
+expression:
+
+<<< Discount.re#catch-exception{5}
+
+With the current logic, it's not impossible to reach the exception branch, but
+who knows how the code might evolve in the future.
```

Comment:
Maybe this could be shared as an alternative approach? Rather than "as an extra precaution", something like "as an alternative approach", and then share the code with exn handling and without the `.length` check.

Reply:
Agree, catching the exception is described as alternate approach in https://github.com/melange-re/melange-for-react-devs/pull/34/commits/4184366e2ce38b16b59e39e0739c64dc363877fc